src/api/context/mod.rs

```diff
@@ -0,0 +1,379 @@
+//!
+//! [`Context`]s are immutable, and their write operations result in the creation
+//! of a new context containing the original values and the new specified values.
+//!
+//! ## Context state
+//!
+//! Concerns can create and retrieve their local state in the current execution
+//! state represented by a context through the [`get`] and [`with_value`]
+//! methods. It is recommended to use application-specific types when storing new
+//! context values to avoid unintentionally overwriting existing state.
+//!
+//! ## Managing the current context
+//!
+//! Contexts can be associated with the caller's current execution unit on a
+//! given thread via the [`attach`] method, and previous contexts can be restored
+//! by dropping the returned [`ContextGuard`]. Context can be nested, and will
+//! restore their parent outer context when detached on drop. To access the
+//! values of the context, a snapshot can be created via the [`Context::current`]
+//! method.
+//!
+//! [`Context`]: struct.Context.html
+//! [`Context::current`]: struct.Context.html#method.current
+//! [`ContextGuard`]: struct.ContextGuard.html
+//! [`get`]: struct.Context.html#method.get
+//! [`with_value`]: struct.Context.html#method.with_value
+//! [`attach`]: struct.Context.html#method.attach
+//!
+//! # Examples
+//!
+//! ```
+//! use opentelemetry::api::Context;
+//!
+//! // Application-specific `a` and `b` values
+//! #[derive(Debug, PartialEq)]
+//! struct ValueA(&'static str);
+//! #[derive(Debug, PartialEq)]
+//! struct ValueB(u64);
+//!
+//! let _outer_guard = Context::new().with_value(ValueA("a")).attach();
+//!
+//! // Only value a has been set
+//! let current = Context::current();
+//! assert_eq!(current.get::<ValueA>(), Some(&ValueA("a")));
+//! assert_eq!(current.get::<ValueB>(), None);
+//!
+//! {
+//!     let _inner_guard = Context::current_with_value(ValueB(42)).attach();
+//!     // Both values are set in inner context
+//!     let current = Context::current();
+//!     assert_eq!(current.get::<ValueA>(), Some(&ValueA("a")));
+//!     assert_eq!(current.get::<ValueB>(), Some(&ValueB(42)));
+//! }
+//!
+//! // Resets to only the `a` value when inner guard is dropped
+//! let current = Context::current();
+//! assert_eq!(current.get::<ValueA>(), Some(&ValueA("a")));
+//! assert_eq!(current.get::<ValueB>(), None);
+//! ```
+
+use std::any::{Any, TypeId};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::fmt;
+use std::hash::{BuildHasherDefault, Hasher};
+use std::rc::Rc;
+
+thread_local! {
+    static CURRENT_CONTEXT: RefCell<Context> = RefCell::new(Context::default());
+    static DEFAULT_CONTEXT: Context = Context::default();
+}
+
+/// An execution-scoped collection of values.
+#[derive(Clone, Default)]
+pub struct Context {
+    entries: HashMap<TypeId, Rc<dyn Any>, BuildHasherDefault<IdHasher>>,
+}
+
+impl Context {
+    /// Creates an empty `Context`.
+    ///
+    /// The context is initially created with a capacity of 0, so it will not
```

Comment:
Does this mean the default context is not usable? Does this require the application / user to create a context?

I believe the intention is for the context to always be available and useable.

Reply:
This context is still usable, it just doesn't heap allocate, basically just the same comment from https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.new